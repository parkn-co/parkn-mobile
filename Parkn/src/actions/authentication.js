import * as AuthApi from '../api/authentication';

export const AUTH_IS_FETCHING = 'AUTH_IS_FETCHING';
export function setIsFetching() {
  return {
    type: AUTH_IS_FETCHING,
  };
}

export const AUTH_SET_FORM_VALUES = 'AUTH_SET_FORM_VALUES';
export function setFormValues(values) {
  return {
    type: AUTH_SET_FORM_VALUES,
    payload: values,
  };
}

export const AUTH_SIGN_IN = 'AUTH_SIGN_IN';
export const AUTH_SET_ERRORS = 'AUTH_SET_ERRORS';
export function authenticateWithValues(isSignUp, values) {
  return (dispatch, getState) => {
    dispatch({
      type: AUTH_SET_FORM_VALUES,
      payload: values
    });

    dispatch(setIsFetching());

    const {
      authentication: {form}
    } = getState();

    let method = 'signIn';
    if (isSignUp) {
      method = 'signUp';
    }

    AuthApi[method]({...form, ...values})
    .then(token => dispatch({
      type: AUTH_SIGN_IN,
      payload: token,
    }))
    .catch(err => {
      let payload = err;

      if (err.status === 409) {
        payload = {...err, email: err.error};
      } else if (err.status === 404 || err.status === 500) {
        payload = {...err, email: 'Incorrect email or password'};
      }

      dispatch({
        type: AUTH_SET_ERRORS,
        payload,
      })
    });
  }
}

export const AUTH_FETCH_USER = 'AUTH_FETCH_USER';
export function fetchUser() {
  return (dispatch, getState) => {
    dispatch(setIsFetching());

    AuthApi.fetchUser(getState)
    .then(token => dispatch({
      type: AUTH_FETCH_USER,
      payload: token,
    }))
    .catch(err => dispatch({
      type: AUTH_SET_ERRORS,
      payload: err,
    }));
  }
}
