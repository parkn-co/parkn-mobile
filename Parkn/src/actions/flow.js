export type Action = {
  type: String,
  payload: ?Any
};

export type AsyncAction = (dispatch: Function, getState: Function) => void
