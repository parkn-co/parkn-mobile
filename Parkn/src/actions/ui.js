// @flow
import type {Action} from 'flow-declarations/redux';

export const SWITCH_TO_VIEW = 'SWITCH_TO_VIEW';
export const switchToView = (payload: Object): Action => ({
  type: SWITCH_TO_VIEW,
  payload,
});
