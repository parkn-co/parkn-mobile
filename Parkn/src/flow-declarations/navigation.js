// @flow
import React, {Component} from 'react';

export type Route = {
  id: string,
  component: Component<*, *, *>,
  nav?: Object,
  isInitial?: boolean,
  keys?: Array<string>,
  isSignUp?: boolean,
};

export type NavButtonGenerator = (
  route: Route,
  navigator: any,
  index: number,
  navState: any
) => React.Element<*>;
