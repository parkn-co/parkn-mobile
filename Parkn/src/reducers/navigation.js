// @flow
import type {Action} from 'flow-declarations/redux';
// Import action handlers
import {
  NAV_NAVIGATE_TO,
  NAV_DID_NAVIGATE_TO,
  NAV_SET_IS_NAVIGATING,
} from '../actions/navigation';

import {REHYDRATE} from 'redux-persist/constants';

// Action Handlers
const REDUCER_ACTION_HANDLERS = {
   [NAV_NAVIGATE_TO]: (state: Object, {payload}: Action) => ({
     ...state,
     navigateToRoute: payload,
   }),
   [NAV_DID_NAVIGATE_TO]: (state: Object, {payload}: Action) => ({
     ...state,
     navigateToRoute: null,
     route: payload,
     isNavigating: false,
   }),
   [NAV_SET_IS_NAVIGATING]: (state: Object, {payload}: Action) => ({
     ...state,
     isNavigating: payload,
   }),
   [REHYDRATE]: () => initialState,
};

const initialState = {
  navigateToRoute: null,
  route: null,
  isNavigating: false,
};

// Reducer
export default function navigationReducer(state: Object = initialState, action: Action) {
  const handler = REDUCER_ACTION_HANDLERS[action.type];
  return handler ? handler(state, action) : state;
}
