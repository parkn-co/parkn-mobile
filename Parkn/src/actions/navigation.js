// @flow
import {Action} from './flow';

export const NAV_NAVIGATE_TO = 'NAVIGATE_TO';
export function navigateTo(route: Object): Action {
  return {
    type: NAV_NAVIGATE_TO,
    payload: route,
  };
}

export const NAV_DID_NAVIGATE_TO = 'DID_NAVIGATE_TO';
export function didNavigateTo(route: Object): Action {
  return {
    type: NAV_DID_NAVIGATE_TO,
    payload: route,
  }
}

export const NAV_SET_IS_NAVIGATING = 'NAV_SET_IS_NAVIGATING';
export function setIsNavigating(isNavigating: Boolean): Action {
  return {
    type: NAV_SET_IS_NAVIGATING,
    payload: isNavigating,
  }
}
