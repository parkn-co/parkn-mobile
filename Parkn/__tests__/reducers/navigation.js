/* eslint-env jest */

import React from 'react';
import { REHYDRATE } from 'redux-persist/constants';
import * as navigationActions from '../../src/actions/navigation';
import navigationReducer from '../../src/reducers/navigation';

describe('Navigation reducer', () => {
  const blankState = {
    navigateToRoute: null,
    route: null,
    isNavigating: false,
  };
  const route = {
    id: 'Landing',
    component: jest.fn(),
  };

  it('returns the correct new state when NAV_NAVIGATE_TO is dispatched', () => {
    const action = {
      type: navigationActions.NAV_NAVIGATE_TO,
      payload: route,
    };
    expect(navigationReducer(blankState, action)).toMatchSnapshot();
  });

  it('returns the correct new state when NAV_DID_NAVIGATE_TO is dispatched', () => {
    const action = {
      type: navigationActions.NAV_DID_NAVIGATE_TO,
      payload: route,
    };
    expect(navigationReducer(blankState, action)).toMatchSnapshot();
  });

  it('returns the correct new state when NAV_SET_IS_NAVIGATING is dispatched', () => {
    const action = {
      type: navigationActions.NAV_SET_IS_NAVIGATING,
      payload: true,
    };
    expect(navigationReducer(blankState, action)).toMatchSnapshot();
  });

  it('returns the correct new state when REHYDRATE is dispatched', () => {
    const action = {
      type: navigationActions.REHYDRATE,
    };
    expect(navigationReducer(blankState, action)).toMatchSnapshot();
  });
});
