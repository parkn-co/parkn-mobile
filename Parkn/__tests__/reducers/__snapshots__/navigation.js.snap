exports[`Navigation reducer returns the correct new state when NAV_DID_NAVIGATE_TO is dispatched 1`] = `
Object {
  "isNavigating": false,
  "navigateToRoute": null,
  "route": Object {
    "component": [Function],
    "id": "Landing",
  },
}
`;

exports[`Navigation reducer returns the correct new state when NAV_NAVIGATE_TO is dispatched 1`] = `
Object {
  "isNavigating": false,
  "navigateToRoute": Object {
    "component": [Function],
    "id": "Landing",
  },
  "route": null,
}
`;

exports[`Navigation reducer returns the correct new state when NAV_SET_IS_NAVIGATING is dispatched 1`] = `
Object {
  "isNavigating": true,
  "navigateToRoute": null,
  "route": null,
}
`;

exports[`Navigation reducer returns the correct new state when REHYDRATE is dispatched 1`] = `
Object {
  "isNavigating": false,
  "navigateToRoute": null,
  "route": null,
}
`;
